docs/proposed/http-storage-node-protocol.rst

```diff
@@ -0,0 +1,506 @@
+The request body is the raw share data (i.e., ``application/octet-stream``).
+*Content-Range* requests are encouraged for large transfers.
+For example,
+for a 1MiB share the data can be broken in to 8 128KiB chunks.
+Each chunk can be *PUT* separately with the appropriate *Content-Range* header.
+The server must recognize when all of the data has been received and mark the share as complete
+(which it can do because it was informed of the size when the storage index was initialized).
+Clients should upload chunks in re-assembly order.
+Servers may reject out-of-order chunks for implementation simplicity.
+If an individual *PUT* fails then only a limited amount of effort is wasted on the necessary retry.
+
+.. think about copying https://developers.google.com/drive/api/v2/resumable-upload
+
+``POST /v1/immutable/:storage_index/:share_number/corrupt``
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+Advise the server the data read from the indicated share was corrupt.
+The request body includes an human-meaningful string with details about the corruption.
+It also includes potentially important details about the share.
+
+For example::
+
+  {"reason": "expected hash abcd, got hash efgh"}
+
+.. share-type, storage-index, and share-number are inferred from the URL
+
+Reading
+~~~~~~~
+
+``GET /v1/immutable/:storage_index/shares``
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+Retrieve a list indicating all shares available for the indicated storage index.
+For example::
+
+  [1, 5]
+
+``GET /v1/immutable/:storage_index?share=:s0&share=:sN&offset=o1&size=z0&offset=oN&size=zN``
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+Read data from the indicated immutable shares.
+If ``share`` query parameters are given, selecte only those shares for reading.
+Otherwise, select all shares present.
+If ``size`` and ``offset`` query parameters are given,
+only the portions thus identified of the selected shares are returned.
+Otherwise, all data is from the selected shares is returned.
+
+The response body contains a mapping giving the read data.
+For example::
+
+  {
+      3: ["foo", "bar"],
+      7: ["baz", "quux"]
+  }
+
+Discussion
+``````````
+
+Offset and size of the requested data are specified here as query arguments.
+Instead, this information could be present in a ``Range`` header in the request.
+This is the more obvious choice and leverages an HTTP feature built for exactly this use-case.
+However, HTTP requires that the ``Content-Type`` of the response to "range requests" be ``multipart/...``.
+The ``multipart`` major type brings along string sentinel delimiting as a means to frame the different response parts.
+There are many drawbacks to this framing technique:
+
+1. It is resource-intensive to generate.
+2. It is resource-intensive to parse.
+3. It is complex to parse safely [#]_ [#]_ [#]_ [#]_.
+
+Mutable
+-------
+
+Writing
+~~~~~~~
+
+``POST /v1/mutable/:storage_index``
```

Comment:
This operation is *apparently* unnecessary.  With the Foolscap-based protocol, clients just start writing to a share of a storage index - they don't allocate buckets first.  Perhaps this makes sense as mutable shares can vary in size whereas immutable shares have a fixed size assigned at bucket allocation time.

I'm not completely convinced.  However, for GBS v1, it may make the most sense to mirror the Foolscap application-level protocol as closely as possible to minimize contentiousness/complexity/etc.  So I'm considering dropping this endpoint.

Reply:
On the plus side, it's going to be way easier to make v2 *add* an endpoint rather than go through a "deprecate, then drop entirely" type of dance .. so especially if it doesn't appear *necessary* now, I'm +1 on dropping it too...